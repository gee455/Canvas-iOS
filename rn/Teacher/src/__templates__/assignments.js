//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

/* @flow */

import template, { type Template } from '../utils/template'

export const assignment: Template<Assignment> = template({
  id: '123456789',
  course_id: '987654321',
  name: 'learn to write code',
  description: 'a course that can take any beginner to professional in a matter of microseconds.  Harness the power of your favorite languages like PHP and Cold Fusion.',
  due_at: '2037-06-01T05:59:00Z',
  unlock_at: '2017-03-17T06:00:00Z',
  lock_at: '2037-06-01T05:59:59Z',
  created_at: '2017-03-17T19:15:25Z',
  updated_at: '2017-03-17T19:15:25Z',
  points_possible: 20,
  grading_type: 'points',
  assignment_group_id: 1376,
  peer_reviews: false,
  automatic_peer_reviews: false,
  position: 1,
  submission_types:
  [ 'online_text_entry',
    'online_url',
    'media_recording',
    'online_upload' ],
  needs_grading_count: 0,
  html_url: 'http://hostname.tld/whatever',
  overrides: [],
  grade_group_students_individually: false,
  group_category_id: null,
  has_overrides: false,
  only_visible_to_overrides: false,
  published: true,
  unpublishable: false,
  rubric_assessment: {},
  rubric_settings: null,
  rubric: null,
})

export const assignmentGroup: Template<AssignmentGroup> = template({
  id: '1',
  name: 'Learn React Native',
  position: 1,
  assignments: [assignment()],
  group_weight: 1,
  integration_data: {},
  sis_source_id: '',
})

export const assignmentDueDate: Template<AssignmentDate> = template({
  title: 'due',
  due_at: '2037-06-01T05:59:00Z',
  unlock_at: '2017-03-17T06:00:00Z',
  lock_at: '2037-06-01T05:59:59Z',
})

export const assignmentOverride: Template<AssignmentOverride> = template({
  id: '1',
  assignment_id: '1',
  student_ids: [],
  group_id: '1',
  course_section_id: '1',
  title: 'override',
  all_day: false,
  all_date_date: '',
  due_at: '2037-06-01T05:59:00Z',
  unlock_at: '2017-03-17T06:00:00Z',
  lock_at: '2037-06-01T05:59:59Z',
})
