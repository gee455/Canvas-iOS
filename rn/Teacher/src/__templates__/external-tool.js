//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

/* @flow */

import template, { type Template } from '../utils/template'

export const externalTool: Template<AddressBookResult> = template({
  id: '1',
  name: 'LTI FTW',
  url: 'https://canvas.instructure.com/lti',
})

export const ltiLaunchDefinition: Template<LtiLaunchDefinition> = template({
  definition_id: 42,
  placements: {
    course_navigation: {
      url: 'https://rollcall.instructure.com/launchy-launch',
    },
  },
})
