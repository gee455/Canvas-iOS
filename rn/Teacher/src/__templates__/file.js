//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

/* @flow */

import template, { type Template } from '../utils/template'

export const usageRights: Template<UsageRights> = template({
  legal_copyright: '',
  use_justification: 'own_copyright',
})

export const file: Template<File> = template({
  id: '111',
  display_name: 'Book Report',
  url: 'http://canvaslms.com/bookreport',
  size: 488212,
  thumbnail_url: 'http://fillmurray.com/322/200',
  mime_class: 'document',
  preview_url: '/',
  filename: '/bookreport',
  parent_folder_id: null,
  locked: false,
  hidden: false,
  unlock_at: null,
  lock_at: null,
  usage_rights: usageRights(),
  'content-type': 'application/document',
})
