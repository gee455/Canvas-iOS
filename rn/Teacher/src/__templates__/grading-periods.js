//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

/* @flow */

import template, { type Template } from '../utils/template'

export const gradingPeriod: Template<GradingPeriod> = template({
  id: '1023',
  title: 'First Block',
  start_date: '2014-01-07T15:04:00Z',
  end_date: '2014-05-07T17:07:00Z',
  close_date: '2014-06-07T17:07:00Z',
  weight: 33.33,
})
