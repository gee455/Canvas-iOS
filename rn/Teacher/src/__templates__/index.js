//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// @flow

export * from '../redux/__templates__/app-state'
export * from './account-notification'
export * from './addressBook'
export * from './api'
export * from './assignments'
export * from './attachment'
export * from './conversations'
export * from './course'
export * from './discussion'
export * from './enrollments'
export * from './error'
export * from './external-tool'
export * from './file'
export * from './folder'
export * from './grading-periods'
export * from './group'
export * from './helm'
export * from './launch-definitions'
export * from './mediaComment'
export * from './page'
export * from './quiz'
export * from './quizSubmission'
export * from './role'
export * from './rubric'
export * from './section'
export * from './session'
export * from './submissions'
export * from './tab'
export * from './toDo'
export * from './users'
