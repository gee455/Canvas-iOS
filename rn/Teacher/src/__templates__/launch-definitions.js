//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// @flow

import template, { type Template } from '../utils/template'

export const launchDefinitionGlobalNavigationItem: Template<ExternalToolLaunchDefinitionGlobalNavigationItem> = template({
  message_type: 'basic-lti-launch-request',
  url: 'https://mobiledev.beta.instructuremedia.com/lti/launch?custom_gauge_launch_type=global_nav',
  title: 'Gauge',
})

export const launchDefinition: Template<ExternalToolLaunchDefinition> = template({
  definition_type: 'ContextExternalTool',
  definition_id: '360860',
  name: 'Gauge',
  description: 'Tools for Education',
  domain: 'gauge.instructure.com',
  placements: {
    global_navigation: launchDefinitionGlobalNavigationItem(),
  },
})
