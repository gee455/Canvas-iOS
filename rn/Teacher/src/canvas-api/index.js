//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// @flow

import 'babel-polyfill'

export { default } from './apis/index'
export * from './apis/assignmentGroups'
export * from './apis/assignments'
export * from './apis/conversations'
export * from './apis/courses'
export * from './apis/discussions'
export * from './apis/enrollments'
export * from './apis/external-tools'
export * from './apis/groups'
export * from './apis/login'
export * from './apis/quizzes'
export * from './apis/submissions'
export * from './apis/users'
export * from './apis/media-comments'
export * from './apis/file-uploads'
export * from './apis/files'
export * from './apis/accounts'

export * from './session'

export { default as httpClient, isAbort, httpCache } from './httpClient'
