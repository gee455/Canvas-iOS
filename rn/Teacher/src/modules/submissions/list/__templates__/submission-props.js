//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// @flow

import template, { type Template } from '../../../../utils/template'
import type { SubmissionDataProps } from '../submission-prop-types'

const templates = {
  ...require('../../../../__templates__/submissions'),
}

const submissionWithHistory = templates.submissionHistory()

// $FlowFixMe
export const submissionProps: Template<SubmissionDataProps> = template({
  grade: 'B-',
  name: submissionWithHistory.user.name,
  status: 'submitted',
  userID: submissionWithHistory.user.id,
  avatarURL: submissionWithHistory.user.avatar_url,
  sectionID: '1',
})
