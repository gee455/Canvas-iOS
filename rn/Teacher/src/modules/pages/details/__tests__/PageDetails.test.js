//
// Copyright (C) 2016-present Instructure, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

// @flow

import { shallow } from 'enzyme'
import React from 'react'
import { Alert, ActionSheetIOS } from 'react-native'
import URL from 'url-parse'
import { alertError } from '../../../../redux/middleware/error-handler'
import { API, httpCache } from '../../../../canvas-api/model-api'
import * as template from '../../../../__templates__'
import app from '../../../app'
import ConnectedPageDetails, { PageDetails } from '../PageDetails'

jest.mock('../../../../redux/middleware/error-handler', () => {
  return { alertError: jest.fn() }
})
jest.useFakeTimers()

describe('PageDetails', () => {
  let props
  beforeEach(() => {
    httpCache.clear()
    app.setCurrentApp('teacher')
    props = {
      location: new URL('/pages/page-1#jumpto'),
      courseID: '1',
      url: 'page-1',
      navigator: template.navigator(),
      page: template.pageModel({ url: 'page-1' }),
      course: template.courseModel({ id: '1', name: 'Course 1' }),
      courseColor: '#fff',
      api: new API({ policy: 'cache-only' }),
      isLoading: false,
      loadError: null,
      refresh: jest.fn(),
    }
  })

  it('gets courseColor, course, and page from the model api', () => {
    const courseColor = 'green'
    const course = [ template.courseModel() ]
    const page = template.pageModel({ url: 'test' })
    httpCache.handle('GET', 'users/self/colors', { custom_colors: { course_1: courseColor } })
    httpCache.handle('GET', 'courses/1', course)
    httpCache.handle('GET', 'courses/1/pages/test', page)
    const tree = shallow(<ConnectedPageDetails courseID='1' url='test' />)
    expect(tree.find(PageDetails).props()).toMatchObject({
      courseColor,
      course,
      page,
    })
  })

  it('gets front page if the url matches', () => {
    const page = template.pageModel({ url: 'somewhere' })
    httpCache.handle('GET', 'users/self/colors', { custom_colors: { course_1: 'green' } })
    httpCache.handle('GET', 'courses/1', template.courseModel())
    httpCache.handle('GET', 'courses/1/front_page', page)
    const tree = shallow(<ConnectedPageDetails courseID='1' url='front_page' />)
    expect(tree.find(PageDetails).props()).toMatchObject({
      page,
    })
  })

  it('renders', () => {
    const tree = shallow(<PageDetails {...props} />)
    expect(tree).toMatchSnapshot()
  })

  it('renders while loading', async () => {
    props.course = null
    props.page = null
    props.isLoading = true
    const tree = shallow(<PageDetails {...props} />)
    expect(tree).toMatchSnapshot()
  })

  it('alerts on new loadError', () => {
    const tree = shallow(<PageDetails {...props} />)
    tree.setProps({ loadError: null })
    expect(alertError).not.toHaveBeenCalled()
    const loadError = new Error()
    tree.setProps({ loadError })
    expect(alertError).toHaveBeenCalledWith(loadError)
  })

  it('routes to page edit', async () => {
    // $FlowFixMe
    ActionSheetIOS.showActionSheetWithOptions = jest.fn((options, callback) => callback(0))
    props.navigator = template.navigator({ show: jest.fn() })
    props.courseID = '1'
    props.url = 'page-1'
    const tree = shallow(<PageDetails {...props} />)
    tree.find('Screen').prop('rightBarButtons')
      .find(({ testID }) => testID === 'pages.details.editButton')
      .action()
    expect(props.navigator.show).toHaveBeenCalledWith(
      '/courses/1/pages/page-1/edit',
      { modal: true, modalPresentationStyle: 'formsheet' },
      { onChange: expect.any(Function) }
    )
  })

  it('deletes page', async () => {
    // $FlowFixMe
    ActionSheetIOS.showActionSheetWithOptions = jest.fn((options, callback) => callback(1))
    // $FlowFixMe
    Alert.alert = jest.fn((title, message, buttons) => buttons[1].onPress())

    props.api.deletePage = jest.fn()
    props.page = template.pageModel({ url: 'page-1', isFrontPage: false })
    const tree = shallow(<PageDetails {...props} />)
    tree.find('Screen').prop('rightBarButtons')
      .find(({ testID }) => testID === 'pages.details.editButton')
      .action()
    expect(props.api.deletePage).toHaveBeenCalledWith('courses', props.courseID, 'page-1')
  })

  it('cant delete front page', async () => {
    // $FlowFixMe
    ActionSheetIOS.showActionSheetWithOptions = jest.fn((options, callback) => callback(1))
    // $FlowFixMe
    Alert.alert = jest.fn((title, message, buttons) => buttons[1].onPress())

    props.api.deletePage = jest.fn()
    props.page = template.pageModel({ url: 'page-1', isFrontPage: true })
    const tree = shallow(<PageDetails {...props} />)
    tree.find('Screen').prop('rightBarButtons')
      .find(({ testID }) => testID === 'pages.details.editButton')
      .action()
    expect(props.api.deletePage).not.toHaveBeenCalled()
  })

  it('alerts error deleting page', async () => {
    // $FlowFixMe
    ActionSheetIOS.showActionSheetWithOptions = jest.fn((options, callback) => callback(1))
    // $FlowFixMe
    Alert.alert = jest.fn((title, message, buttons) => buttons[1].onPress())

    const error = new Error()
    const rejection = Promise.reject(error)
    props.api.deletePage = jest.fn(() => rejection)
    props.page = template.pageModel({ url: 'page-1', isFrontPage: false })
    const tree = shallow(<PageDetails {...props} />)
    tree.find('Screen').prop('rightBarButtons')
      .find(({ testID }) => testID === 'pages.details.editButton')
      .action()
    expect(props.api.deletePage).toHaveBeenCalledWith('courses', props.courseID, 'page-1')
    await rejection.catch(() => {})
    expect(alertError).toHaveBeenCalledWith(error)
  })

  it('shows edit button if permitted', () => {
    app.setCurrentApp('teacher')
    const tree = shallow(<PageDetails {...props} />)
    const button = tree.find('Screen').prop('rightBarButtons')
      .find(({ testID }) => testID === 'pages.details.editButton')
    expect(button).toBeDefined()
  })

  it('hides edit button if not permitted', () => {
    app.setCurrentApp('student')
    const tree = shallow(<PageDetails {...props} />)
    expect(tree.find('Screen').prop('rightBarButtons')).toBeFalsy()
  })

  it('bails out without error when page is null', () => {
    props.page = null
    const tree = shallow(<PageDetails {...props} />)
    expect(tree.instance().edit).not.toThrow()
    expect(tree.instance().showEditActionSheet).not.toThrow()
    expect(tree.instance()._editActionSheetSelected).not.toThrow()
    expect(tree.instance().delete).not.toThrow()
  })

  it('replaces navigation when edited page url changes', () => {
    props.navigator = template.navigator({
      show: jest.fn(),
      replace: jest.fn(),
    })
    const tree = shallow(<PageDetails {...props} />)
    tree.instance().edit()
    const { onChange } = props.navigator.show.mock.calls[0][2]
    onChange(props.page)
    expect(props.navigator.replace).not.toHaveBeenCalled()
    onChange(template.pageModel({ url: 'updated-3' }))
    expect(props.navigator.replace).toHaveBeenCalledWith(
      '/courses/1/pages/updated-3'
    )
  })
})
