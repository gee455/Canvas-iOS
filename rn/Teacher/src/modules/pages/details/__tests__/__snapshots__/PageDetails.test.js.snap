// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageDetails renders 1`] = `
<Screen
  navBarColor="#fff"
  navBarStyle="dark"
  rightBarButtons={
    Array [
      Object {
        "accessibilityLabel": "Options",
        "action": [Function],
        "image": Object {
          "testUri": "../../../src/images/kabob.png",
        },
        "testID": "pages.details.editButton",
      },
    ]
  }
  showDismissButton={true}
  statusBarStyle="default"
  subtitle="Course 1"
  title="Page Details"
>
  <RefreshableScrollView
    onRefresh={[Function]}
    refreshing={false}
    style={
      Object {
        "flex": 1,
        "padding": 16,
      }
    }
  >
    <Heading1
      style={
        Object {
          "paddingBottom": 17,
        }
      }
    >
      Page 1
    </Heading1>
    <CanvasWebView
      automaticallySetHeight={true}
      navigator={
        Object {
          "dismiss": [Function],
          "dismissAllModals": [Function],
          "isModal": false,
          "pop": [Function],
          "show": [Function],
          "traitCollection": [Function],
        }
      }
      scrollEnabled={true}
      source={
        Object {
          "baseURL": "/pages/page-1#jumpto",
          "html": "<p>Hello, World!</p>",
        }
      }
      style={
        Object {
          "flex": 1,
        }
      }
    />
  </RefreshableScrollView>
</Screen>
`;

exports[`PageDetails renders while loading 1`] = `
<Screen
  navBarColor="#fff"
  navBarStyle="dark"
  rightBarButtons={
    Array [
      Object {
        "accessibilityLabel": "Options",
        "action": [Function],
        "image": Object {
          "testUri": "../../../src/images/kabob.png",
        },
        "testID": "pages.details.editButton",
      },
    ]
  }
  showDismissButton={true}
  statusBarStyle="default"
  title="Page Details"
>
  <RefreshableScrollView
    onRefresh={[Function]}
    refreshing={true}
    style={
      Object {
        "flex": 1,
        "padding": 16,
      }
    }
  >
    <Heading1
      style={
        Object {
          "paddingBottom": 17,
        }
      }
    />
    <CanvasWebView
      automaticallySetHeight={true}
      navigator={
        Object {
          "dismiss": [Function],
          "dismissAllModals": [Function],
          "isModal": false,
          "pop": [Function],
          "show": [Function],
          "traitCollection": [Function],
        }
      }
      scrollEnabled={true}
      source={
        Object {
          "baseURL": "",
          "html": "",
        }
      }
      style={
        Object {
          "flex": 1,
        }
      }
    />
  </RefreshableScrollView>
</Screen>
`;
